plugin/trino-apache-ranger/src/main/java/io/trino/plugin/ranger/RangerSystemAccessControl.java

```diff
@@ -0,0 +1,1097 @@
+        }
+    }
+
+    private Instant getQueryTime(SystemSecurityContext context)
+    {
+        if (context != null) {
+            return getQueryTime(context.getQueryId());
+        }
+        else {
+            return null;
+        }
+    }
+
+    private boolean hasPermission(RangerTrinoResource resource, SystemSecurityContext context, RangerTrinoAccessType accessType, String action)
+    {
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(createAccessRequest(resource, context, accessType, action));
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private boolean hasPermissionForFilter(RangerTrinoResource resource, SystemSecurityContext context, RangerTrinoAccessType accessType, String action)
+    {
+        RangerTrinoAccessRequest request = createAccessRequest(resource, context, accessType, action);
+
+        request.setResourceMatchingScope(RangerAccessRequest.ResourceMatchingScope.SELF_OR_DESCENDANTS);
+
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(request, null);
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private boolean hasPermission(RangerTrinoResource resource, Identity identity, QueryId queryId, RangerTrinoAccessType accessType, String action)
+    {
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(createAccessRequest(resource, identity, queryId, accessType, action));
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private boolean hasPermissionForFilter(RangerTrinoResource resource, Identity identity, QueryId queryId, RangerTrinoAccessType accessType, String action)
+    {
+        RangerTrinoAccessRequest request = createAccessRequest(resource, identity, queryId, accessType, action);
+
+        request.setResourceMatchingScope(RangerAccessRequest.ResourceMatchingScope.SELF_OR_DESCENDANTS);
+
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(request, null);
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private boolean hasPermission(RangerTrinoResource resource, Optional<Principal> principal, QueryId queryId, RangerTrinoAccessType accessType, String action)
+    {
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(createAccessRequest(resource, toIdentity(principal), queryId, accessType, action));
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private static RangerTrinoResource createTableResource(CatalogSchemaTableName catalogSchemaTableName)
+    {
+        return RangerTrinoResource.forTable(catalogSchemaTableName.getCatalogName(), catalogSchemaTableName.getSchemaTableName().getSchemaName(), catalogSchemaTableName.getSchemaTableName().getTableName());
+    }
+
+    private static Identity toIdentity(Optional<Principal> principal)
+    {
+        if (principal.isPresent()) {
+            return Identity.ofUser(principal.get().getName());
+        }
+        else {
+            return Identity.ofUser("");
```

Comment:
It does not look correct to create such identities? What is the use case of that?

Reply:
Following method sends `Optional<Principal>` - which doesn't seem correct. Perhaps the authorizer should ignore such cases?

```
    @Deprecated
    @Override
    public void checkCanSetUser(Optional<Principal> principal, String userName)
    {
        if (!hasPermission(RangerTrinoResource.forUser(userName), principal, null, IMPERSONATE, "SetUser")) {
            denySetUser(principal, userName);
        }
    }

````